subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java

```diff
@@ -209,4 +210,22 @@ public int hashCode() {
         result = 31 * result + asImmutable().hashCode();
         return result;
     }
+
+    private <T> void addLazyAttribute(Attribute<T> key, Provider<? extends T> provider) {
+        final Map<Attribute<?>, Provider<?>> temp = Maps.newHashMap(lazyAttributes);
+        temp.put(key, provider);
+        lazyAttributes = ImmutableMap.copyOf(temp);
+    }
+
+    private <T> Provider<? extends T> removeLazyAttribute(Attribute<T> key) {
+        @SuppressWarnings("unchecked")
+        final Provider<? extends T> result = (Provider<? extends T>) lazyAttributes.get(key);
+
+        final Map<Attribute<?>, Provider<?>> temp = Maps.newHashMap(lazyAttributes);
+        //noinspection SuspiciousMethodCalls
+        temp.remove(lazyAttributes);
+        lazyAttributes = ImmutableMap.copyOf(temp);
+
+        return result;
+    }
```

Comment:
Do we really get anything out of using an immutable map here?  We end up just creating a bunch of new maps vs updating a single instance?

Reply:
Just trying to use a different `Map` implementation to minimize storage, and was influenced by the pattern used for storing state.  I'll switch to a pre-sized `LinkedHashMap`.

Reply:
> Do we really get anything out of using an immutable map here?

We always get good stuff from using immutable data structures. In this particular case:
* no `null` checks anywhere, no branching
* maps can be immediately shared without copying

> We end up just creating a bunch of new maps vs updating a single instance?

Lots of assumptions here :)

Consider, for instance, that CHAMP data structures have lower memory overhead than JDK collections.

These Guava ImmutableMap with copying behavior provide some nice properties but are, in general, much less sophisticated implementations of these concepts.

While I understand the reluctance to applying CHAMP here (lack of familiarity) I still think it's the superior general solution.

Reply:
> ...  I think we could do something simple here with an empty map sentinel value vs using null. ...

Sounds good to me 👍